staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go

```diff
@@ -108,15 +109,81 @@ func newETCD3Client(c storagebackend.Config) (*clientv3.Client, error) {
 	return client, err
 }
 
+type runningCompactor struct {
+	interval time.Duration
+	cancel   context.CancelFunc
+	client   *clientv3.Client
+	refs     int
+}
+
+var (
+	lock       sync.Mutex
+	compactors = map[string]*runningCompactor{}
+)
+
+// startCompactorOnce start one compactor per transport. If the interval get smaller on repeated calls, the
+// compactor is replaced. A destroy func is returned. If all destroy funcs with the same transport are called,
+// the compactor is stopped.
+func startCompactorOnce(c storagebackend.TransportConfig, interval time.Duration) (func(), error) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	key := fmt.Sprintf("%v", c)
+	if compactor, ok := compactors[key]; !ok || compactor.interval > interval {
+		if ok {
+			// replace compactor
+			compactor.cancel()
+			compactor.client.Close()
+		} else {
+			// start new compactor
+			compactor = &runningCompactor{}
+		}
+
+		compactorClient, err := newETCD3Client(c)
+		if err != nil {
+			return nil, err
+		}
+
+		ctx, cancel := context.WithCancel(context.Background())
+
+		compactor.interval = interval
+		compactor.cancel = cancel
+		compactor.client = compactorClient
+		compactor.refs++
+		compactors[key] = compactor
+
+		etcd3.StartCompactor(ctx, compactorClient, interval)
+	} else {
+		compactor.refs++
+	}
+
+	return func() {
+		lock.Lock()
+		defer lock.Unlock()
+
+		compactor := compactors[key]
+		compactor.refs--
+		if compactors[key].refs == 0 {
+			compactor.cancel()
+			compactor.client.Close()
+			delete(compactors, key)
```

Comment:
deleting by key here this looks dangerous. we may delete wrong storage on racing by mistake amiright? deleting by pointer equality would be safer.

Reply:
not sure I can follow. The compactor is not by storage, but by transport config. Moreover, we lock before reading or writing to `compactors`.